talisker/testing.py

```diff
@@ -131,7 +131,7 @@ def cmp(a, b):
                 return False
 
         if all(cmp(getattr(record, k, _s), v) for k, v in kwargs.items()):
-            if extra:
+            if extra and hasattr(record, 'extra'):
```

Comment:
It'd be nice to add a regression test for this. Like

```python
def test_log_record_list_no_extra():
    makeRecord = logging.Logger('test').makeRecord

    def record(name, level, msg, extra=None):
        return makeRecord(name, level, 'fn', 123, msg, None, None)

    r1 = record('root.log1', logging.WARNING, 'foo')
    r2 = record('root.log2', logging.INFO, 'foo bar')
    records = testing.LogRecordList()
    records.extend([r1, r2])

    assert records.filter(name='root.log1') == [r1]
    assert records.filter(msg='foo') == [r1, r2]
    assert records.filter(level=logging.WARNING) == [r1]
    assert records.filter(extra={'a': 1}) == []
```

in tests/test_testing.py.

And this test fails the last assertion because the fix above results in `filter(extra=...)` matching all records that do not have `.extra`. So I think we should return False here if `extra` is passed in `filter()` but the record doesn't have `.extra`. What do you think?